### PR TITLE
adding a disk size stats for the index

### PR DESCRIPTION
**Pull request by sreekanth-cb**
The new "num_bytes_used_disk_by_root" stats shows
the disk size used by the latest root index.
The idea is that this should help in suggesting
a sizing recommendation for RAM quota driven by
an index size resident ratio.

### CODE REVIEW

File: index/scorch/scorch.go [hunk 1]
```diff
@@ -507,8 +533,9 @@ func (s *Scorch) StatsMap() map[string]interface{} {
 	m["num_items_introduced"] = m["TotIntroducedItems"]
 	m["num_items_persisted"] = m["TotPersistedItems"]
 	m["num_recs_to_persist"] = m["TotItemsToPersist"]
-	m["num_bytes_used_disk"] = m["CurOnDiskBytes"]
-	m["num_files_on_disk"] = m["CurOnDiskFiles"]
+	m["num_bytes_used_disk"] = numBytesUsedDisk
```
Comment (mschoch) [inline index/scorch/scorch.go]: According to the comments in this section, some of the stats have 2 names for backwards compatibility.  But if that is the case, I don't think we should keep introducing multiple names for each stat.  Or is there more to this than I understand?

File: index/scorch/scorch.go [hunk 2]
```diff
@@ -466,33 +466,59 @@ func (s *Scorch) Stats() json.Marshaler {
...
+	s.rootLock.RLock()
+	rootSegPaths := s.rootDiskSegmentsPaths()
+	m["CurFilesIneligibleForRemoval"] = uint64(len(s.ineligibleForRemoval))
+	s.rootLock.RUnlock()
+
+	numFilesOnDisk, numBytesUsedDisk, numBytesOnDiskByRoot := s.diskFileStats(rootSegPaths)
 
 	m["CurOnDiskBytes"] = numBytesUsedDisk
 	m["CurOnDiskFiles"] = numFilesOnDisk
+	m["CurOnDiskBytesByRoot"] = numBytesOnDiskByRoot
```
Comment (mschoch) [inline index/scorch/scorch.go]: Can you add a description for each of these new stats to the commit message?  Just like a sentence is enough.  The problem is that the stat name uses "root" which we use inside of scorch to mean a particular thing, but that isn't obvious to people outside of the scorch internals.

File: index/scorch/scorch.go [hunk 3]
```diff
@@ -466,33 +466,59 @@ func (s *Scorch) Stats() json.Marshaler {
...
+						if _, fileAtRoot := rootSegmentPaths[fname]; fileAtRoot {
+							numBytesOnDiskByRoot += uint64(finfo.Size())
+						}
+					}
 				}
 			}
 		}
 	}
-	return numFilesOnDisk, numBytesUsedDisk
+	// if no root files path given, then consider all disk files.
```
Comment (mschoch) [inline index/scorch/scorch.go]: is this actually needed?  the persister pause logic passes nil, but only looks at the number of files on disk.

it just seems like it's adding complexity, defining special behavior for nil, but that behavior isn't used presently.

File: index/scorch/scorch.go (hunk 1 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: agree..i was just following the pattern there. So m["CurOnDiskBytesByRoot"]  isn't needed here.

File: index/scorch/scorch.go (hunk 2 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: CurOnDiskBytesByIndex or CurOnDiskIndexBytes ?  
Currently we don't have a stat which explicitly gives the index size on disk.  CurOnDiskBytes computes the whole files size in the index directory which might include the older snapshots as well.
Exact index disk size helps to suggest a resident ratio/ram quota for the index to the customer - is the context for this change,

File: index/scorch/scorch.go (hunk 3 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: currently its not used, but felt this helps to have a more consistent behaviour for that function.
like - anyone calling with nil/users not bothered about the exact index disk size, should get the whole file size in directory than just receiving zero as a return value for the 3rd/new return value.

File: index/scorch/scorch.go [hunk 4]
```diff
@@ -507,8 +532,11 @@ func (s *Scorch) StatsMap() map[string]interface{} {
 	m["num_items_introduced"] = m["TotIntroducedItems"]
 	m["num_items_persisted"] = m["TotPersistedItems"]
 	m["num_recs_to_persist"] = m["TotItemsToPersist"]
-	m["num_bytes_used_disk"] = m["CurOnDiskBytes"]
-	m["num_files_on_disk"] = m["CurOnDiskFiles"]
+	// total disk bytes found in index directory inclusive of older snapshots
+	m["num_bytes_used_disk"] = numBytesUsedDisk
+	// total disk bytes by index exclusive of older snapshots
+	m["num_bytes_used_disk_index"] = numBytesOnDiskByIndex
```
Comment (abhinavdangeti) [inline index/scorch/scorch.go]: This stat seems more like "num_bytes_used_disk" to me.
The older num_bytes_used_disk can be "num_bytes_on_disk" - what do you think?

Comment (sreekanth-cb) [reply index/scorch/scorch.go]: older stats name could be something like a "total_num_bytes_on_disk" etc, but we can't change stats due to backward compatibility as we don't know whether anyone actually started using it or not ?

**Comment by sreekanth-cb** [comment]
@mschoch ,  can you please review this.

File: index/scorch/scorch.go (hunk 2 above)
Comment (mschoch) [reply index/scorch/scorch.go]: So, my request was for you to add it to the commit message, as I think that people are more likely to consult that if they have questions about it.

Second, I disagree with the statement 'we don't have a stat which explicitly gives the index size on disk'.  If you have chosen to keep more than 1 snapshot, or you have queries using older snapshots, then your index does contain multiple snapshots.  So, the size of your index is the size of all of them.

That doesn't mean there isn't value in knowing the size of just the current root.  We just disagree on the definitions.

File: index/scorch/scorch.go (hunk 1 above)
Comment (mschoch) [reply index/scorch/scorch.go]: OK, so it will be removed?

File: index/scorch/scorch.go (hunk 3 above)
Comment (mschoch) [reply index/scorch/scorch.go]: OK, I don't feel too strongly about this, just dislike writing code that isn't used.

File: index/scorch/scorch.go (hunk 4 above)
Comment (mschoch) [reply index/scorch/scorch.go]: Yes, we certainly cannot rename stat keys as they are just strings passed through to users.

**Comment by mschoch** [comment]
@sreekanth-cb I reviewed and replied to your comments, but I don't see any changes since the last review.

File: index/scorch/scorch.go (hunk 1 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: which one to be removed now? m["CurOnDiskBytesByRoot"]  is removed already if you thats what you were referring to.

File: index/scorch/scorch.go (hunk 2 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: do you have a better naming suggestion as we disagree here on the definitions?

**Comment by sreekanth-cb** [comment]
@mschoch I did make changes, but did force amend on the commit.

File: index/scorch/scorch.go (hunk 2 above)
Comment (sreekanth-cb) [reply index/scorch/scorch.go]: ok, let me go back to the older CurOnDiskBytesByRoot name itself your request was about just changing the commit msg.

Comment (mschoch) [reply index/scorch/scorch.go]: i sense that you don't agree about something, but i don't ow what that is

Comment (sreekanth-cb) [reply index/scorch/scorch.go]: i just misread your original comment about renaming the stat from root -> index, but then on a second read I realised that its more about the commit description than about the exact naming stuff..

**Pull request merged**